File: crates/benches/fixtures/config_loader.go
package config

import (
	"errors"
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

// ServerConfig holds HTTP server settings.
type ServerConfig struct {
	Host    string `yaml:"host"`
	Port    int    `yaml:"port"`
	Workers int    `yaml:"workers"`
	TLS     bool   `yaml:"tls"`
}

// DatabaseConfig holds database connection settings.
type DatabaseConfig struct {
	URL            string `yaml:"url"`
	MaxConnections int    `yaml:"max_connections"`
	IdleTimeoutSec int    `yaml:"idle_timeout_sec"`
}

// AppConfig is the top-level configuration struct.
type AppConfig struct {
	Server   ServerConfig   `yaml:"server"`
	Database DatabaseConfig `yaml:"database"`
	LogLevel string         `yaml:"log_level"`
	Features map[string]bool `yaml:"features"`
}

// DefaultConfig returns a config with sensible defaults.
func DefaultConfig() *AppConfig {
	return &AppConfig{
		Server:   ServerConfig{Host: "0.0.0.0", Port: 8080, Workers: 4},
		LogLevel: "info",
		Features: make(map[string]bool),
	}
}

// LoadFile loads configuration from a YAML file.
func LoadFile(path string) (*AppConfig, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open config: %w", err)
	}
	defer f.Close()
	cfg := DefaultConfig()
	if err := yaml.NewDecoder(f).Decode(cfg); err != nil {
		return nil, fmt.Errorf("decode config: %w", err)
	}
	return cfg, cfg.Validate()
}

// Validate checks required fields.
func (c *AppConfig) Validate() error {
	if c.Database.URL == "" {
		return errors.New("database.url is required")
	}
	if c.Server.Port <= 0 {
		return fmt.Errorf("server.port must be > 0, got %d", c.Server.Port)
	}
	return nil
}

// MergeEnv overrides config values from environment variables.
func MergeEnv(cfg *AppConfig) {
	if url := os.Getenv("DATABASE_URL"); url != "" {
		cfg.Database.URL = url
	}
	if level := os.Getenv("LOG_LEVEL"); level != "" {
		cfg.LogLevel = level
	}
}

File: crates/benches/fixtures/config_loader.rs
use std::collections::HashMap;
use std::env;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ConfigError {
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
    #[error("Parse error: {0}")]
    Parse(String),
    #[error("Validation error: {field} — {message}")]
    Validation { field: String, message: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ServerConfig {
    pub host: String,
    pub port: u16,
    pub workers: usize,
    pub tls: bool,
}

impl Default for ServerConfig {
    fn default() -> Self {
        Self { host: "127.0.0.1".into(), port: 8080, workers: 4, tls: false }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseConfig {
    pub url: String,
    pub max_connections: u32,
    pub idle_timeout_secs: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppConfig {
    pub server: ServerConfig,
    pub database: DatabaseConfig,
    pub log_level: String,
    pub features: HashMap<String, bool>,
}

impl AppConfig {
    pub fn from_file(path: &Path) -> Result<Self, ConfigError> {
        let content = std::fs::read_to_string(path)?;
        let cfg: Self = toml::from_str(&content).map_err(|e| ConfigError::Parse(e.to_string()))?;
        cfg.validate()?;
        Ok(cfg)
    }

    pub fn from_env() -> Result<Self, ConfigError> {
        let url = env::var("DATABASE_URL").unwrap_or_default();
        Ok(Self {
            server: ServerConfig::default(),
            database: DatabaseConfig { url, max_connections: 10, idle_timeout_secs: 30 },
            log_level: env::var("LOG_LEVEL").unwrap_or_else(|_| "info".into()),
            features: HashMap::new(),
        })
    }

    fn validate(&self) -> Result<(), ConfigError> {
        if self.database.url.is_empty() {
            return Err(ConfigError::Validation { field: "database.url".into(), message: "must not be empty".into() });
        }
        if self.server.port == 0 {
            return Err(ConfigError::Validation { field: "server.port".into(), message: "must be > 0".into() });
        }
        Ok(())
    }
}
